src/services/suggestionDiagnostics.ts

```diff
@@ -109,7 +113,118 @@ namespace ts {
         }
     }
 
+    function addConvertToAsyncFunctionDiagnostics(node: Node, checker: TypeChecker, diags: DiagnosticWithLocation[]): void {
+        if (isAsyncFunction(node)) {
+            return;
+        }
+
+        if (isFunctionLikeDeclaration(node) && !node.body) {
+            return; // break on ambient functions
+        }
+
+        const returnType = checker.getTypeAtLocation(node);

```

Comment:
i am not sure this returns the return type, this will return the type of the function itself.. i think you want getReturnTypeOfSignature on the return type instead of using it directly.

Reply:
I debugged through this locally, and i see the wrong type, i also see the check failing. i think the tests for some reason do not have  the lib included, and thus `Promise` is not defined, and an error type is used instead. see `getGlobalPromiseType` implementation for more details. this results in the `getGlobalPromiseType` being assigned to anything.

I think the correct thing to do here is to expose `getPromisedTypeOfPromise` on checker, and use that instead. if it returns `undefined` you know the original type is not a `Promise`.

Also you need to get the call signatures, then get the return type of that to check if it is a promise or not.. e.g.:

```ts
        const functionType = checker.getTypeAtLocation(node);
        const callSignatures = checker.getSignaturesOfType(functionType, SignatureKind.Call);
        const returnType = callSignatures.length && checker.getReturnTypeOfSignature(callSignatures[0]);
        if (!returnType || !checker.getPromisedTypeOfPromise(returnType)) {
            return;
        }
```